### ISSUE TITLE
Http request is ignoring timeout if it can connect to the server

### ISSUE THREAD

**Issue by dested** [eclipsesource/tabris-js#890]
It seems in the event the xmlhttprequest connects to the server, it is timing out after 10 seconds, instead of honoring the timeout property passed in. It throws a java.net.SocketTimeoutException as shown in the picture. 

http://i.imgur.com/L2xYsCV.png

**Comment by mpost** [comment]
I presume you are talking about Android here. The connection timeout is supported. Could you provide a snippet where the problem occurs?

**Comment by dested** [comment]
Yes, android. 

```var xhr = new XMLHttpRequest();
                xhr.open('GET', someSlowUrl);
                xhr.timeout = 100*1000;
                xhr.ontimeout = (e) => {
                    console.log("Timeout");
                };
                xhr.send();
                ```

The issue is the connection is established but taking too long, as opposed to it not being able to initiate the request. If I change it to ```someUnfoundUrl``` it times out for the appropriate amount of time. 

In my naive searching I believe it is related to setting ```httpRequest.setSoTimeout``` as well as ```httpRequest.setConnectionTimeout```, but I could be wrong.

**Comment by mpost** [comment]
You are right in your observation... the `xhr.timeout` affects the connection timeout not the read or write timeout. The default for the read and write timeout is 10 sec each.

**Comment by dested** [comment]
Is this a feature that would be possible to expose? 

It seems when I execute this via the web the xhr.timeout variable reflects both values, as the connection does not timeout in the same way.

**Comment by mpost** [comment]
@dested Yes i agree. The spec for the timeout states that it does not affect one or the other but the entire duration. In our underlying implementation we can specify the three timeouts but they do not (yet) effect each other. We should look into to distribute the timeout time evenly between all three timeouts.

**Comment by mpost** [comment]
Btw... the fetch() standard does not have a timeout at all and requires the user to use custom promises to simulate the behavior.

**Comment by dested** [comment]
Did this make it into 1.9?

**Comment by mpost** [comment]
@dested No change on that front. iOS and Android both offer more fine grained control on the timeouts than what the XHR offers. We are currently evaluating if it makes more sense to offer these properties on the XHR on top of the original spec.

**Comment by dested** [comment]
I'd like to put this back on the radar if possible. The arbitrary ten second timeout is a sincere issue. If nothing else it would be nice to see it increased to something more reasonable like 30-60 seconds. My fear is http requests that take eleven seconds due for one reason or another are coming back as failed, and forcing my users to think they were not executed, thus causing them to be executed multiple times. 

Are there any current thoughts on implementation?

**Comment by cookieguru** [comment]
@dested Can you use fetch as Moritz suggested?

**Comment by dested** [comment]
Very interesting, I did not realize these capabilities in fetch. I have retooled to use that in my application and I am now able to manage my state much saner. Feel free to close this ticket,  but I do think the ten second timeout in XmlHttpRequest should be noted somewhere.

**Comment by cookieguru** [comment]
@dested It's still a valid issue and should be left open as someone might be working with a third party library which needs the support and/or can't convert to fetch.

**Comment by ralfstx** [comment]
The timeout is currently mapped only to the connection timeout. Mapping it to the read timeout as well would extend the timeout.

We could consider setting the native timeouts to infinity and calling the native `abort()` method after the configured timeout from JS.
